src/gs-side-filter-row.c

```diff
@@ -0,0 +1,146 @@
+/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
+ *
+ * Copyright (C) 2016 Endless Mobile, Inc.
+ * Copyright (C) 2013 Matthias Clasen <mclasen@redhat.com>
+ *
+ * Licensed under the GNU General Public License Version 2
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "config.h"
+
+#include <glib/gi18n.h>
+#include <gtk/gtk.h>
+
+#include "gs-side-filter-row.h"
+#include "gs-common.h"
+
+struct _GsSideFilterRow
+{
+	GtkListBoxRow	parent_instance;
+
+	GsCategory	*cat;
+	GsShellMode	 mode;
+	GtkWidget	*label;
+	GtkWidget	*image;
+	GtkWidget	*leftborder;
+};
+
+G_DEFINE_TYPE (GsSideFilterRow, gs_side_filter_row, GTK_TYPE_LIST_BOX_ROW)
+
+GsCategory *
+gs_side_filter_row_get_category (GsSideFilterRow *row)
+{
+	g_return_val_if_fail (GS_IS_SIDE_FILTER_ROW (row), NULL);
+
+	return row->cat;
+}
+
+void
+gs_side_filter_row_set_category (GsSideFilterRow *row, GsCategory *cat)
+{
+	GPtrArray *key_colors;
+
+	g_return_if_fail (GS_IS_SIDE_FILTER_ROW (row));
+	g_return_if_fail (GS_IS_CATEGORY (cat));
+
+	g_clear_object (&row->cat);
+	row->cat = g_object_ref (cat);
```

Comment:
`g_set_object()` can be used without the category being a property, and has the advantage of also working in case `cat` and `row->cat` are exactly the same instance (the current code may release the last reference of the instance when calling `g_clear_object()`)